certbot-ci/certbot_integration_tests/utils/acme_server.py

```diff
@@ -11,176 +11,168 @@
 
         # Wait for the ACME CA server to be up.
-        print('=> Waiting for {0} instance to respond...'.format(acme_type))
-        misc.check_until_timeout(acme_xdist['directory_url'])
+        print('=> Waiting for boulder instance to respond...')
+        misc.check_until_timeout(self.acme_xdist['directory_url'])
 
         # Configure challtestsrv to answer any A record request with ip of the docker host.
-        acme_subnet = '10.77.77' if acme_type == 'boulder' else '10.30.50'
-        response = requests.post('http://localhost:{0}/set-default-ipv4'
-                                 .format(acme_xdist['challtestsrv_port']),
-                                 json={'ip': '{0}.1'.format(acme_subnet)})
+        response = requests.post('http://localhost:{0}/set-default-ipv4'.format(CHALLTESTSRV_PORT),
+                                 json={'ip': '10.77.77.1'})
         response.raise_for_status()
 
-        print('=> Finished {0} instance deployment.'.format(acme_type))
-    except BaseException:
-        print('Error while setting up {0} instance.'.format(acme_type))
-        raise
-
+        print('=> Finished boulder instance deployment.')
 
-def _prepare_http_proxy(acme_xdist):
-    """Configure and launch an HTTP proxy"""
-    print('=> Configuring the HTTP proxy...')
-    mapping = {r'.+\.{0}\.wtf'.format(node): 'http://127.0.0.1:{0}'.format(port)
-               for node, port in acme_xdist['http_port'].items()}
-    command = [sys.executable, proxy.__file__, str(HTTP_01_PORT), json.dumps(mapping)]
-    process = subprocess.Popen(command, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
-    print('=> Finished configuring the HTTP proxy.')
+    def _prepare_http_proxy(self):
+        """Configure and launch an HTTP proxy"""
+        print('=> Configuring the HTTP proxy...')
+        mapping = {r'.+\.{0}\.wtf'.format(node): 'http://127.0.0.1:{0}'.format(port)
+                   for node, port in self.acme_xdist['http_port'].items()}
+        command = [sys.executable, proxy.__file__, str(HTTP_01_PORT), json.dumps(mapping)]
+        self._launch_process(command)
+        print('=> Finished configuring the HTTP proxy.')
 
-    return process
-
-
-def _launch_command(command, cwd=os.getcwd()):
-    """Launch silently an OS command, output will be displayed in case of failure"""
-    try:
-        subprocess.check_output(command, stderr=subprocess.STDOUT, cwd=cwd, universal_newlines=True)
-    except subprocess.CalledProcessError as e:
-        sys.stderr.write(e.output)
-        raise
+    def _launch_process(self, command, cwd=os.getcwd(), env=None):
+        """Launch silently an subprocess OS command"""
+        if not env:
+            env = os.environ
+        with open(os.devnull, 'w') as null:
+            process = subprocess.Popen(command, stdout=null, stderr=subprocess.STDOUT, cwd=cwd, env=env)
+            self._processes.append(process)
+            return process
```

Comment:
Aren't we closing the file descriptor that still may be being used by the process here? I'm open to other approaches, but maybe it makes sense to make a single `/dev/null` file descriptor in the class that can be shared between all processes and closed when `stop()` is called. What do you think?

Reply:
Yes, you are right. Not elegant at all.

In fact I though of this again, and there are two typical cases here, about what you want to do with the output generated by the subprocesses (acme server, http proxy ...):
1) In the context of integration tests using pytest, we want to silence this out. What is interesting is the output of Certbot, not the noise of the third parties, that we assume working correctly
2) In the context of `run_acme_server`, either to test Certbot or troubleshoot these third parties. In this case, we want the stdout of these processes to be displayed in the standard stdout, to be consumed as it is necessary

So, since `ACMEServer` is a context manager with everything needed for setup/teardown the context, I implemented a stdout alongside its lifecycle, to be used by all suprocesses. Its nature depends of a `stdout` parameter given to the `ACMEServer` constructor. If `True`, this stdout is sys.stdout, and we see the processes activity in the standard stdout. If `False`, we use a descriptor on `/dev/null` that is opened/closed accordingly in the context manager at entry and exit context.

For 1), `stdout` is False, for 2) it is `True`.